pkg/kubelet/prober/worker.go

```diff
@@ -316,11 +316,14 @@ func (w *worker) doProbe(ctx context.Context) (keepGoing bool) {
 
 	w.resultsManager.Set(w.containerID, result, w.pod)
 
-	if (w.probeType == liveness || w.probeType == startup) && result == results.Failure {
+	if (w.probeType == liveness && result == results.Failure) || w.probeType == startup {
```

Comment:
I don't understand this. So for startup probe we will stop after the first run? This condition doesn't check the result to be a failure.

Reply:
@SergeyKanzhelev 

Currently, `w.onHold = true` will be set when liveness or startup probe is Failure.
However, for asynchronous reason(I've referred to that in [this PR description](https://github.com/kubernetes/kubernetes/pull/121206)), `w.onHold = true` also should also be set when startup probe is Success. Otherwise, startup probe may be executed redundantly.

Similar discussion has been done in:
https://github.com/kubernetes/kubernetes/pull/121206#discussion_r1366735410
https://github.com/kubernetes/kubernetes/pull/121206#discussion_r1721013512

Reply:
> `w.onHold = true` also should also be set when startup probe is Success.

What I see in code is that we set it unconditionally for startup type. Not when it is a startup AND succeeded. What am I missing here?

Reply:
@SergeyKanzhelev 

> What I see in code is that we set it unconditionally for startup type. Not when it is a startup AND succeeded.

Yes, that's right.
So I've said
> should also be set when startup probe is Success.

Regarding Startup Probe,  `w.onHold = true` should be set in each case:

- `failureThreshold` is reached and last result is `Failure`
- `SuccessThreshold`(Must be `1`) is reached and last result is `Success`

Threshold check will be done just before this like that, and if it has not been reached, it returns true.
https://github.com/kubernetes/kubernetes/blob/aa8f2878a588d80b0dc7960c0b82af1882011115/pkg/kubelet/prober/worker.go#L316-L320

Therefore, at this point, it means that one of the thresholds has been reached, and I think that `w.onHold = true` should be set for the Startup Probe regardless of the result.
(Currently, it is only for the case of Failure, but also for Success.)

Reply:
/lgtm
/approve